go-jni/frontend/handler.go

```diff
@@ -0,0 +1,939 @@
+package frontend
+
+import (
+	"cpg"
+	"fmt"
+	"go/ast"
+	"go/token"
+	"log"
+	"strconv"
+	"strings"
+
+	"tekao.net/jnigi"
+)
+
+func getImportName(spec *ast.ImportSpec) string {
+	if spec.Name != nil {
+		return spec.Name.Name
+	}
+
+	var path = spec.Path.Value[1 : len(spec.Path.Value)-1]
+	var paths = strings.Split(path, "/")
+
+	return paths[len(paths)-1]
+}
+
+func (this *GoLanguageFrontend) HandleFile(fset *token.FileSet, file *ast.File) (tu *cpg.TranslationUnitDeclaration, err error) {
+	tu = cpg.NewTranslationUnitDeclaration(fset, file, "src.go", this.GetCodeFromRawNode(fset, file))
+
+	scope := this.GetScopeManager()
+
+	// reset scope
+	scope.ResetToGlobal((*cpg.Node)(tu))
+
+	// set current TU
+	this.SetCurrentTU(tu)
+
+	/*for _, imprt := range file.Imports {
+		i := cpg.NewIncludeDeclaration(fset, imprt)
+
+		i.SetName(getImportName(imprt))
+		i.SetFilename(imprt.Path.Value[1 : len(imprt.Path.Value)-1])
+
+		err = scope.AddDeclaration((*cpg.Declaration)(i))
+		if err != nil {
+			log.Fatal(err)
+		}
+	}*/
+
+	// create a new namespace declaration, representing the package
+	p := cpg.NewNamespaceDeclaration(fset, nil, file.Name.Name, fmt.Sprintf("package %s", file.Name.Name))
+
+	// enter scope
+	scope.EnterScope((*cpg.Node)(p))
+
+	for _, decl := range file.Decls {
+		var d *cpg.Declaration
+
+		d = this.handleDecl(fset, decl)
+
+		if d != nil {
+			err = scope.AddDeclaration((*cpg.Declaration)(d))
+			if err != nil {
+				log.Fatal(err)
+
+			}
+		}
+	}
+
+	// leave scope
+	scope.LeaveScope((*cpg.Node)(p))
+
+	// add it
+	scope.AddDeclaration((*cpg.Declaration)(p))
+
+	return
+}
+
+func (this *GoLanguageFrontend) handleDecl(fset *token.FileSet, decl ast.Decl) *cpg.Declaration {
+	this.LogDebug("Handling declaration (%T): %+v", decl, decl)
+
+	switch v := decl.(type) {
```

Comment:
Looks like this PR uses explicit dispatching, in the other handlers the dispatching is done by the handler superclass and such switches are not necessary in the construction code.

Reply:
We cannot use the handler super class since the handler system depends on dispatching handler functions based on a raw ast object that exist in the JVM, i.e., the JavaParser and Eclipse CDT objects. There are no equivalent Java objects in this case, so it is not possible, unless we create dummy objects in the JVM representing the Go objects. And that does not make sense in my opinion.